### PR TITLE
[DataGrid] Allow to set in GridExportCsvOptions delimiter

### PR DESCRIPTION
**Pull request by michallukowski**
Add option in GridExportCsvOptions to set delimiter character

Fix 1. of #1440

### CODE REVIEW

**Comment by dtassone** [comment]
FYI the comma as decimal separator is according to the culture. So in the uk they use 1, 000 and in Fr, they use 1 000 I think.
So not sure we should add it as an option. Maybe we could think of an excel formatter as @m4theushw mentioned somewhere (trying to recover his comment)

**Comment by michallukowski** [comment]
> 
> 
> FYI the comma as decimal separator is according to the culture. So in the uk they use 1, 000 and in Fr, they use 1 000 I think.
> So not sure we should add it as an option. Maybe we could think of an export formatter as @m4theushw mentioned somewhere (trying to recover his comment)

You're talking about the thousand separator. Using Number.prototype.toString() we don't get the thousands separator which I think we shouldn't use in CSV. For example, in many European countries, the number 3.14 in the csv file is automatically treated by Excel as a string. If the decimal separator is ',' (3.14), Excel detects as a number. 
[https://en.wikipedia.org/wiki/Decimal_separator](https://en.wikipedia.org/wiki/Decimal_separator)

File: packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts [hunk 1]
```diff
@@ -6,24 +6,43 @@ import {
   GridRowModel,
 } from '../../../../models';
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:8]: No blank lines in the import

```suggestion
```

File: packages/grid/_modules_/grid/models/gridExport.ts [hunk 2]
```diff
@@ -1,8 +1,34 @@
...
+ */
+export type CsvDelimiterCharacter = ',' | ';';
+
 /**
  * The options to apply on the CSV export.
  */
 export interface GridExportCsvOptions {
+  /**
+   * If `true`, the comma character used to separate the integer part from the fractional part of a number.
+   * This flag also changes the delimiter from `,` to `;`.
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/gridExport.ts:13]: typo
```suggestion
   * This flag also changes the delimiter from `,` to `.`.
```

**Comment by m4theushw** [comment]
@michallukowski I proposed in https://github.com/mui-org/material-ui-x/issues/1814#issuecomment-852678630 that we apply the value formatter to the CSV export too. With it, you would be able to change how numbers are formatted when exported. This is what Damien refers to. Regarding this PR, I would avoid adding specific options to control number formatting. We could leverage [`Intl.NumberFormat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat) that already provides everything. We could ask the user to pass a `Intl.NumberFormat` instance and use it to format values.

> Using Number.prototype.toString() we don't get the thousands separator which I think we shouldn't use in CSV.

`Number.prototype.toLocaleString()` gives the formatted value according to the current locale.

File: packages/grid/_modules_/grid/models/gridExport.ts [hunk 3]
```diff
@@ -1,8 +1,34 @@
...
+   * @default false
+   */
+  commaAsDecimalSeparator?: boolean;
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/gridExport.ts:16]: Could we remove this option and move it to a standalone PR? It seems to create friction for another change this we could merge as-is.

**Review by oliviertassinari — COMMENTED** [review]
Can we add a test case for the new `delimiter` param? 

> Using Number.prototype.toString() we don't get the thousands separator which I think we shouldn't use in CSV.

Agree, I don't see why we would need to format a number in CSV export, it seems better to be able to rely on the CSV reader to automatically detect a number and handle it accordingly (e.g. for formatting based on the locale of the CSV reader or further math operations).

**Comment by dtassone** [comment]
Yes so basically don't do anything special for numbers, we should consider using formatters on export

File: packages/grid/_modules_/grid/models/gridExport.ts (hunk 2 above)
Comment (michallukowski) [reply packages/grid/_modules_/grid/models/gridExport.ts:13]: It's not a typo. When decimal separator is ',' then delimiter shouldn't also be ',' so it is automatically changed to ';'.

File: packages/grid/_modules_/grid/models/gridExport.ts (hunk 3 above)
Comment (michallukowski) [reply packages/grid/_modules_/grid/models/gridExport.ts:16]: Ok. I'll get rid off this feature in this PR.

**Comment by DanailH** [comment]
> Can we add a test case for the new `delimiter` param?
> 
> > Using Number.prototype.toString() we don't get the thousands separator which I think we shouldn't use in CSV.
> 
> Agree, I don't see **why** we would want to format a number in CSV exports. It seems better to be able to rely on the CSV reader to automatically detect a number and handle it accordingly (e.g. for formatting based on the locale of the CSV reader or further math operations).

This reminds me that I don't think that the CSV export currently uses the `valueFormatter` altho it does use `valueGetter`.

File: packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts [hunk 4]
```diff
@@ -31,13 +31,16 @@ export function serialiseRow(
...
 }
 
-export function buildCSV(
-  columns: GridColumns,
-  rows: Map<GridRowId, GridRowModel>,
-  selectedRows: Record<string, GridRowId>,
-  getCellValue: (id: GridRowId, field: string) => GridCellValue,
-  delimiterCharacter: CsvDelimiterCharacter = ',',
-): string {
+type BuildCSVOptions = {
```
Comment (dtassone) [inline packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:34]: It should be an interface

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:34]: @dtassone Ah, right, my bad. Could you push this as a new commit directly on the PR so we can move forward with it?

**Review by oliviertassinari — APPROVED** [review]
